### ISSUE TITLE
Off-screen CustomPaint widgets consume a lot of memory

### ISSUE THREAD

**Issue by ravenblackx** [flutter/flutter#86357]
## Details

1.  Please tell us exactly how to reproduce the problem you are running into, and how you measured the performance.

Instantiate a widget something like
```
SingleChildScrollWidget(
  child: List.generate(500, (_) => Container(
    height: 200,
    child: CustomPaint(
      painter: Sky(),
    ),
  ),
)
```
(Using `class Sky` from https://api.flutter.dev/flutter/rendering/CustomPainter-class.html which doesn't seem to have any reason to be memory-hungry.)

Observe the memory usage, either in observatory or in the fact that a low-memory device crashes.

Contrast
```
SingleChildScrollWidget(
  child: List.generate(500, (_) => Container(
    height: 200,
    decoration: BoxDecoration(shape: BoxShape.circle, color: Colors.red),
  ),
)
```

Whose memory usage is not nearly so dramatic.

The high memory usage and crash occurs in profile mode on a physical device. I'm not sure if we can switch to master channel in our environment though.

In the actual test case, the CustomPaint widgets were only about 80x80, and 300 of them caused memory usage to go up by around 100MB.

The same memory consumption occurs if the CustomPaint widgets are explicitly in an Offstage widget.

**Target Platform:** (not sure if I'm allowed to say)
**Target OS version/browser:** (not sure if I'm allowed to say)
**Devices:** (not sure if I'm allowed to say)

Flutter doctor output partially redacted:
```
[redacted]
[  +38 ms] [✓] Flutter (Channel google3, on Debian GNU/Linux rodete
5.10.40-1rodete1-amd64,
               locale en_US.UTF-8)
[   +7 ms]     • Framework revision bc1b91c4b3 (5 days ago),
2021-07-08T00:00:00.000
[   +4 ms]     • Engine revision c989ae17b7
[   +4 ms]     • Dart version e4125ef075
[redacted]
[  +31 ms] [✓] Android toolchain - develop for Android devices (Android SDK
version Stable)
[   +4 ms]     • Android SDK at google3
[   +4 ms]     • Platform Stable, build-tools Stable
[   +4 ms]     • Java binary at: /usr/local/buildtools/java/jdk/bin/java
[   +5 ms]     • Java version OpenJDK Runtime Environment (build
                 11.0.11+9-google-release-376318671)
[   +6 ms] [!] Android Studio (not installed)
[   +4 ms]     • Android Studio not found; download from
                 https://developer.android.com/studio/index.html
[   +4 ms]       (or visit
https://flutter.dev/docs/get-started/install/linux#android-setup
                 for detailed instructions).
[  +17 ms] List of devices attached
[   +8 ms] [✓] Connected device (1 available)
[   +4 ms]     • Linux (desktop) [redacted]
[   +5 ms] [✓] Google3 (on linux)
[   +3 ms]     • KVM enabled
[   +4 ms] ! Doctor found issues in 1 category.
[  +19 ms] "flutter doctor" took 554ms.
[+4333 ms] ensureAnalyticsSent: 4318ms
[  +22 ms] exiting with code 0
```

I realize "use a slivered scroller instead of a SingleChildScrollView" is an obvious workaround for *this specific oversimplified example*, but the real use-case needs to measure the heights of all the widgets (which can be impacted by word-wrap in their internal elements), which it seems you can't do without fully instantiating them.

**Comment by maheshmnj** [comment]
HI @ravenblackx,
Please provide a [minimal complete reproducible](https://stackoverflow.com/help/minimal-reproducible-example) code, along with the complete output of ```flutter doctor -v```, this helps to better investigate the issue.

Thank you.

**Comment by ravenblackx** [comment]
Is there some way to make memory usage issues recognizably reproducible? I searched for any kind of way to retrieve the current memory usage from the VM so I could demonstrate this with a test, but I had a hard time finding any way to examine memory usage other than pushing a profile build to an actual device, connecting the observatory, and inspecting the memory usage manually, which is a horrible experience for attempting to diagnose or repro memory issues.

https://github.com/flutter/flutter/issues/21585#issuecomment-879848690 summarizes my findings of trying to find a way to do this.

**Comment by maheshmnj** [comment]
@ravenblackx, You could try using the [dev tools](https://flutter.dev/docs/development/tools/devtools/overview).

**Comment by ravenblackx** [comment]
From trying that in as much as I can, I think it's probably the case that the memory it uses is VRAM. I can't confirm this because I can't find any way to monitor VRAM usage; I'm extrapolating from the fact that it shows up as using a lot of memory on the device that I believe has very little VRAM and then uses shared RAM, and I can't detect any memory usage increase on other devices.

Unfortunately, I can't just push arbitrary flutter binaries to the peculiar device, so I can't make a smaller repro in the one place where I can demonstrably detect the behavior, and if I run a smaller repro on any *other* device, the problem isn't demonstrable because every other device I have available has VRAM so it doesn't use the measurable memory. (I presume - it also could be something else like maybe different cache decisions are made on these devices.)

For more context about peculiar devices, ref. Google-internal b/192756648

**Comment by maheshmnj** [comment]
@ravenblackx 
The sample code you provided in the description isn't valid, You need to provide the correct sample code along with the unmodified output of ```flutter doctor -v``` to address the issue.

**Comment by ravenblackx** [comment]
A sample that I *think* would repro the behavior (but I can't tell because there doesn't seem to be any way to monitor VRAM allocation):

```
import 'package:flutter/material.dart';

final decoratedBoxes = List.generate(
  500,
  (_) => Container(
    height: 200,
    width: 200,
    decoration: BoxDecoration(shape: BoxShape.circle, color: Color(0xffff0000)),
  ),
);

final customPainters = List.generate(
  500,
  (_) => Container(
    height: 200,
    width: 200,
    child: CustomPaint(
      painter: _Blurp(),
    ),
  ),
);

final generatedChildren = ValueNotifier(decoratedBoxes);

Widget testWidget() {
  return Column(
    children: [
      Row(
        children: [
          RaisedButton(
            onPressed: () {
              generatedChildren.value = customPainters;
            },
            child: Text('CustomPainters'),
          ),
          RaisedButton(
            onPressed: () {
              generatedChildren.value = decoratedBoxes;
            },
            child: Text('DecoratedBoxes'),
          ),
        ],
      ),
      Expanded(
        child: SingleChildScrollView(
          child: ValueListenableBuilder<List<Widget>>(
            valueListenable: generatedChildren,
            builder: (_, generatedChildren, __) => Column(
              children: generatedChildren,
            ),
          ),
        ),
      ),
    ],
  );
}

void main() => runApp(StandaloneApp());

class StandaloneApp extends StatelessWidget {
  @override
  Widget build(_) => MaterialApp(
        home: Scaffold(
          body: testWidget(),
        ),
      );
}

class _Blurp extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    final rect = Offset.zero & size;
    const gradient = RadialGradient(
      center: Alignment(0.7, -0.6),
      radius: 0.2,
      colors: <Color>[Color(0xFFFFFF00), Color(0xFF0099FF)],
      stops: <double>[0.4, 1.0],
    );
    canvas.drawRect(
      rect,
      Paint()..shader = gradient.createShader(rect),
    );
  }

  @override
  bool shouldRepaint(_Blurp oldDelegate) => false;
}
```

I can't give unredacted `flutter doctor -v` output because it contains things I don't think I'm allowed to say externally.

**Comment by dnfield** [comment]
This is related to https://github.com/flutter/flutter/issues/12319

Customer wants to be able to scroll to an index and accurately draw a scrollbar.

I'm curious if we could somehow avoid work in `RenderFlex` for this use case.

**Comment by dnfield** [comment]
I have a speculative patch here that seems to avoid painting when a widget is not in the viewport here: https://github.com/flutter/flutter/compare/dnfield:viewport_paint.diff

It breaks some tests - at least some of them are tests that are asserting things painted when they're not in the viewport. It's not entirely clear to me how big of a break it would be to stop doing that. I'm mainly curious if @ravenblackx can patch this in locally and see if it actually makes any difference vram wise.

**Comment by dnfield** [comment]
It has a couple caveats: it does paint widgets that are in the cache extent of the viewport, and it may fail to paint a widget that paints outside of its own bounds until the bounds come into the viewport. This patch is not currently suitable even to evaluate, and it may never be suitable to land, but it might help create a different kind of widget or viewport that does this kind of thing.

**Comment by ravenblackx** [comment]
Yes, with that patch the view that previously went sluggish and then crashed neither goes sluggish nor crashes.

**Comment by goderbauer** [comment]
> I realize "use a slivered scroller instead of a SingleChildScrollView" is an obvious workaround for this specific oversimplified example, but the real use-case needs to measure the heights of all the widgets (which can be impacted by word-wrap in their internal elements), which it seems you can't do without fully instantiating them.

In the real world case, does the extent of each item fully depend on the content of the item and its extent can only be determined by laying each individual item out? Or are there a handful of item classes and all items in a class will have the same extent (which is either known statically upfront or could be determined by laying out one prototype example for each class)? In the latter case, this could also be resolved with a slightly more powerful SliverFixedExtentList that lets you specify different static extents for different indexes. (It could potentially be combined with a PrototypeList, where the extent of each class is measured by laying one example item out instead of all of them).

**Comment by ravenblackx** [comment]
Yeah, as mentioned, potential text-wrapping is involved, so there's no way to do it without either building them all or some kind of off-render text metrics function (which I don't think flutter exposes one of)
